src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -497,6 +478,154 @@ private void StartWriting(byte[] buffer, int offset, int count, AsyncProtocolReq
             }
         }
 
+        // Fill the buffer up to the minimum specified size (or more, if possible).
+        // Returns 0 if EOF on initial read, otherwise throws on EOF.
+        // Returns minSize on success.
+        private int FillBuffer(int minSize)
+        {
+            Debug.Assert(_internalOffset == 0);
+            Debug.Assert(minSize > _internalBufferCount);
+
+            int initialCount = _internalBufferCount;
+            do
+            {
+                int bytes = _sslState.InnerStream.Read(InternalBuffer, _internalBufferCount, InternalBuffer.Length - _internalBufferCount);
+                if (bytes == 0)
+                {
+                    if (_internalBufferCount != initialCount)
+                    {
+                        // We read some bytes, but not as many as we expected, so throw.
+                        throw new IOException(SR.net_io_eof);
+                    }
+
+                    return 0;
+                }
+
+                _internalBufferCount += bytes;
+            } while (_internalBufferCount < minSize);
+
+            return minSize;
+        }
+
+        // Fill the buffer up to the minimum specified size (or more, if possible).
+        // Returns 0 if EOF on initial read, otherwise throws on EOF.
+        // Returns minSize on success.
+        public async Task<int> FillBufferAsync(int minSize)
+        {
+            Debug.Assert(_internalOffset == 0);
+            Debug.Assert(minSize > _internalBufferCount);
+
+            int initialCount = _internalBufferCount;
+            do
+            {
+                int bytes = await _sslState.InnerStream.ReadAsync(InternalBuffer, _internalBufferCount, InternalBuffer.Length - _internalBufferCount, CancellationToken.None).ConfigureAwait(false);
+                if (bytes == 0)
+                {
+                    if (_internalBufferCount != initialCount)
+                    {
+                        // We read some bytes, but not as many as we expected, so throw.
+                        throw new IOException(SR.net_io_eof);
+                    }
+
+                    return 0;
+                }
+
+                _internalBufferCount += bytes;
+            } while (_internalBufferCount < minSize);
+
+            return minSize;
+        }
+
+        private static async void TaskToAsyncProtocolRequest2(Task<int> task, AsyncProtocolRequest asyncRequest)
+        {
+            try
+            {
+                asyncRequest.CompleteRequest(await task.ConfigureAwait(false));
+            }
+            catch (Exception e)
+            {
+                asyncRequest.CompleteUserWithError(e);
+            }
+        }
```

Comment:
I expect this would be a bit more efficient as:
```C#
private static void TaskToAsyncProtocolRequest2(Task<int> task, AsyncProtocolRequest asyncRequest)
{
    if (task.IsCompleted)
    {
        CompleteFromCompletedTask(task, asyncRequest);
    }
    else
    {
        task.ContinueWith((t, s) => CompleteFromCompletedTask(t, (AsyncProtocolRequest)s),
            asyncRequest,
            CancellationToken.None,
            TaskContinuationOptions.ExecuteSynchronously | TaskContinuationOptions.DenyChildAttach,
            TaskScheduler.Default);
    }
}

private static void CompleteFromCompletedTask(Task<int> task, AsyncProtocolRequest asyncRequest)
{
    Debug.Assert(task.IsCompleted);
    if (task.IsRanToCompletion) asyncRequest.CompleteRequest(task.Result);
    else if (task.IsFaulted) asyncRequest.CompleteUserWithError(task.Exception.InnerException);
    else asyncRequest.CompleteUserWithError(new OperationCanceledException());
}
```
The main benefits here are:
- When there's a single await, ContinueWith typically has less overhead, especially if all of the state can be passed via the object state rather than needing to create a heap object to store it.
- For the exceptional case, we're not having to throw it again.